Add CMS management command to update course advanced settings in bulk.

ztraboo opened the pull request:
## Description

Needed a way to set course advanced fields through the Django management command line. Instead of call the REST PATCH endpoint here cms/djangoapps/contentstore/rest_api/v0/views/advanced_settings.py we utilized the `update_course_advanced_settings` method here  cms/djangoapps/contentstore/views/course.py to pass JSON override values for one or more courses.

A `--force-update-catalog-visibility-about` command option was added to include advanced setting. When upgrading from an older release to `Maple` we had this `catalog_visibility` set to `none` which produced the `There was an error loading this course` message when viewing the course MFE Learning frontend. It was necessary to make sure these courses didn't have the `none` value and received `about` or `both` settings to have the page load properly. This command sets the value to `about` to prevent the course from uploading to the `course-discovery` service and out to the marketing frontend.
```
{
    "catalog_visibility": {
        "value": "about"
    }
}
```

## Supporting information

We’re running into this issue mentioned above for students to be able to view the MFE learning page when Course Visibility in Catalog is set to none. The students receive the `There was an error loading this course.` message and nothing loads on the Course Home page. Also the LMS log files indicate no error messages so we were lucky to find this discussion article.

Reference: https://discuss.openedx.org/t/how-to-programmatically-set-a-course-advanced-fields/8901

## Testing instructions

When upgrading from prior release to `Maple` we had several courses with `course_visibility` set to `none` value. With this setting we traversed to the MFE Learning frontend and saw error message with `There was an error loading this course.` and a HTTP 404 response from an EdX API endpoint in the JavaScript console was indicated.

So to regenerate this error just set the `CMS > Advanced Settings > Catalog Visibility` value to `none`.

![image](https://user-images.githubusercontent.com/5641338/207708976-8cb10613-aa01-4770-82fe-9131f097c7fd.png)

HTTP 404 Response was at this address.
https://{{ LMS_HOST }}/api/course_home/course_metadata/course-v1:CUCWD+FAA-ACS-AM-ID-FLF+DOCUMENTATION?browser_timezone=America%2FNew_York

## Invoking Management Command

**WARNING**
We added in a --limit-bulk-update-operation {0-N} command option to prevent the command from consuming all server memory and crashing the server. The default number of courses is set to 150. Please make sure that you pay attention to your server resources before running this command in bulk.

The user passed needs to have superuser rights.

```
# Single Course
# Set the following advanced settings:
# `invitation_only` to `true`
# `catalog_visilibilty` to `about`
python manage.py cms --settings=devstack_docker set_advanced_fields_all_courses \
--fields_json '{"invitation_only": {"value": "true"}}' --user edx@example.com --course course-v1:ORG+COURSENUMBER+COURSERUN --force-update-catalog-visibility-about

# All Courses
# Set the following advanced settings:
# `invitation_only` to `true`
# `catalog_visilibilty` to `about`
# `cerfificate_display_behavior` to `early_no_info`
python manage.py cms --settings=devstack_docker set_advanced_fields_all_courses \
--fields_json '{"invitation_only": {"value": "true"}, "certificates_display_behavior": {"value": "early_no_info"}}' --user edx@example.com --course course-v1:ORG+COURSENUMBER+COURSERUN --force-update-catalog-visibility-about

# All Courses
# Set the following advanced settings:
# `catalog_visilibilty` to `about`
python manage.py cms --settings=devstack_docker set_advanced_fields_all_courses \
--fields_json '{}' --user edx@example.com --course course-v1:ORG+COURSENUMBER+COURSERUN --force-update-catalog-visibility-about
```

ztraboo committed:
feat: Add in CMS management command to set course advanced fields.

Needed a way to set course advanced fields through the Django management command line. Instead of call the REST PATCH endpoint here cms/djangoapps/contentstore/rest_api/v0/views/advanced_settings.py we utilized the `update_course_advanced_settings` method here  cms/djangoapps/contentstore/views/course.py to pass JSON override values for one or more courses.

A `--force-update-catalog-visibility-about` command option was added to include advanced setting. When upgrading from an older release to `Maple` we had this `catalog_visibility` set to `none` which produced the `There was an error loading this course` message when viewing the course MFE Learning frontend. It was necessary to make sure these courses didn't have the `none` value and received `about` or `both` settings to have the page load properly. This command sets the value to `about` to prevent the course from uploading to the `course-discovery` service and out to the marketing frontend.
```
{
    "catalog_visibility": {
        "value": "about"
    }
}
```

Reference: https://discuss.openedx.org/t/how-to-programmatically-set-a-course-advanced-fields/8901

ztraboo committed:
feat: Limit bulk operation for course advanced setting update to avoid memory consumption on server

openedx-webhooks commented:
Thanks for the pull request, @ztraboo!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

mphilbrick211 commented:
Hi @ztraboo! Just checking to see if you are planning to re-run the failing checks on this?

ztraboo committed:
fix: Pylint issues resolved for `set_advanced_fields_all_courses` management command.

Resolved these pylint issues.
```
************* Module cms.djangoapps.contentstore.management.commands.set_advanced_fields_all_courses
cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:117:12: W0707: Consider explicitly re-raising using the 'from' keyword (raise-missing-from)
cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:166:16: R1704: Redefining argument with the local name 'course_id' (redefined-argument-from-local)
cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:222:60: C7630: getattr using a literal attribute name (literal-used-as-attribute)
```

ztraboo commented:
@mphilbrick211 I fixed some pylint issues and rebased on latest master branch.

I noticed this error on some pylint test https://github.com/openedx/edx-platform/actions/runs/3885211137/jobs/6628799798

Failed to install this successfully due to this error.
https://github.com/CUCWD/edx-platform/blob/feature.master/ztraboo/cms-mgmt-set-advanced-fields-all-courses/.github/workflows/pylint-checks.yml#L36-L37
```
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/libg/libgpg-error/libgpg-error-dev_1.37-1_amd64.deb  503  Service Unavailable [IP: 40.119.46.219 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
Error: Process completed with exit code 100.
```

It appears that we're running `sudo apt-get update` prior to `sudo apt-get install libxmlsec1-dev` so wonder if this is something else that others are seeing? Do we need to do this instead `sudo apt-get -y update && sudo apt-get install libxmlsec1-dev`

ztraboo committed:
fix: Fix additional Pylint and Quality test failures for `set_advanced_fields_all_courses`

ztraboo commented:
@mphilbrick211 All test pass now.

mphilbrick211 commented:
Thanks @ztraboo! I will check with my team on having this reviewed.

bradenmacdonald commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+        # WARNING
+        # Limiting the all courses bulk advanced setting update here to avoid consuming all
+        # server memory. Defaulting bulk operations to 150 courses but you can use this argument
+        # to specify any value different than the default.
```
Do you know _why_ this consumes so much memory if the # of courses are not limited? From looking at the code, it looks like it loads each course one at a time, updates its settings, then moves on - so (assuming the garbage collector is working as normal), I would expect that only one course is ever loaded in memory at a time, and there would be no limit to how many courses this can process.

Related: if a server has 10,000 courses but can only process 150 at a time, how can they use this tool to do it? I don't see an option to "skip N courses" so that this could be used in batches.

bradenmacdonald commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+                "value": "about"
+            }
+        }
+        """
```
If you are interested, I think there is a way to make this new command much more generic and remove the specific `force-update-catalog-visibility-about` flag.

You could instead add an `if-fields-json` flag like this:

```
./manage.py cms set_advanced_fields_all_courses \
    --if-fields-json '{"catalog_visibility": "none"}' \
    --fields-json '{"catalog_visibility": "about"}' \
    --user edx@example.com
```

That means "find all courses that have `catalog_visibility: none` and change them to `catalog_visibility: about`".

The nice thing is it's much more generic and flexible this way, _and_ even if this can only be used for ~100 courses at a time, you can just re-run the exact same command as many times as needed until all courses are updated, because it will only "count" courses that match the `if-fields-json` filter.

bradenmacdonald commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+
+    def add_arguments(self, parser):
+        parser.add_argument('--fields_json',
+                            dest='fields_json',
```
I think for consistency, the command-line parameter should use a hyphen: `--fields-json` not `--fields_json`

bradenmacdonald commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+        # Set the following advanced settings:
+        # `invitation_only` to `true`
+        # `catalog_visilibilty` to `about`
+        python manage.py cms --settings=devstack_docker set_advanced_fields_all_courses \
```
I believe `--settings=devstack_docker` is no longer required to be specified, and is definitely wrong on production environments anyways, so should we omit that from this docstring?

bradenmacdonald reviewed (commented):
Hey @ztraboo 👋🏻 , nice work - I think this is a great contribution!

I read through the code and left some questions and comments here - let me know what you think.

One other high-level thought: the way that every parameter requires this extra wrapper of `{"value": ...}` around the value seems a bit inconvenient. Could we update this so that the command inserts that wrapper around the values automatically before passing the field updates on to `update_course_advanced_settings`?

Then the commands would become simpler:

instead of:
```
--fields_json '{"invitation_only": {"value": "true"}, "certificates_display_behavior": {"value": "early_no_info"}}'
```
it could be:
```
--fields-json '{"invitation_only": "true", "certificates_display_behavior": "early_no_info"}'
```

regisb commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+            } \
+        }'
+    user_help = '--user <email> required, needs to be a staff member'
+    course_help = '--course <id> required, e.g. course-v1:ORG+COURSENUMBER+COURSERUN'
```
This argument does not seem to be required, right?

regisb commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+            update_fields,
+            options['user_email'],
+            options['course_id'],
+            int(options['limit_bulk_update_operation'])
```
Argument `limit_bulk_update_operation` should have `type=int` such that we don't have to type-case to int here.

regisb commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+                update_fields, course_module
+            )
+
+            print(f"Applying revised_update_fields = {revised_update_fields} to course {course_id} using `{user}` user.\n")  # pylint: disable=line-too-long
```
Don't use `print`. Instead, use `logger.info` or `logger.debug`. Add a `-v/--verbose` option. Adjust the logger log level based on that option.

regisb commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+            )
+
+            print(f"Applying revised_update_fields = {revised_update_fields} to course {course_id} using `{user}` user.\n")  # pylint: disable=line-too-long
+            print("=" * 80)
```
These line separators are unnecessary.

regisb commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+                    failed_update_advanced_fields_courses.append(str(course_id))
+                    print("=" * 30 + f"> Oops, failed to update fields for {course_id}")
+                    print("Error:")
+                    print(err)
```
Replace these three print statements by: `logger.exception(err)`. I think that by default we should crash on error. Add a `-f/--force` command to ignore errors.

regisb commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+        try:
+            user = user_from_str(user_email)
+        except User.DoesNotExist:
+            logger.warning(user_email + " user does not exist")  # lint-amnesty, pylint: disable=logging-not-lazy
```
We should raise a CommandError here.

regisb commented on cms/djangoapps/contentstore/management/commands/set_advanced_fields_all_courses.py:
```diff
@@ -0,0 +1,248 @@
+            logger.warning("Can't update fields")
+            return
+
+        # Single Course
```
I think that we should handle the single course scenario just like the other one:

```
if course_id:
    course_ids = [course_id]
else:
    course_ids = [str(x.id) for x in modulestore().get_courses()]
for course_id in course_ids[:limit_bulk_update_operation]:
    ...
```

mphilbrick211 commented:
Hi @ztraboo! Just checking in on this to see if this is ready for review/merge?

mphilbrick211 commented:
Hi @ztraboo! Friendly ping on this :)

ztraboo commented:
@mphilbrick211 Sorry I probably need to go back through this and make recommended changes that Braden and Regis noted above before we have another review. I just haven't had time to get to this yet. Please keep this PR open.

mphilbrick211 commented:
> @mphilbrick211 Sorry I probably need to go back through this and make recommended changes that Braden and Regis noted above before we have another review. I just haven't had time to get to this yet. Please keep this PR open.

Hi @ztraboo! just checking in on this :)

mphilbrick211 commented:
One more thing to flag, @ztraboo - two new tests have popped up that need to be run on this pull request (shellcheck). Would you mind running them? Thanks!

mphilbrick211 commented:
> One more thing to flag, @ztraboo - two new tests have popped up that need to be run on this pull request (shellcheck). Would you mind running them? Thanks!

Hi @ztraboo - just following up on my previous comments :)

mphilbrick211 commented:
Hi @ztraboo! Are you planning to pursue this PR? We may need to close it due to inactivity. Please let us know. Thanks!

ztraboo commented:
@mphilbrick211 Sorry for the delay on this. At the moment, I've been swamped with other things. Whenever I can get to this I will.

mphilbrick211 commented:
> @mphilbrick211 Sorry for the delay on this. At the moment, I've been swamped with other things. Whenever I can get to this I will.

Thank you for letting us know!

mphilbrick211 commented:
Hi @ztraboo! I know you have some capacity issues, so this is for when you get back to this...

A few new checks have popped up, including the "shellcheck" tests, as well as the ["pylint" checks](https://discuss.openedx.org/t/minor-change-to-edx-platform-check-statuses/11131). Rebasing should clear these.

e0d committed:
Merge branch 'master' into feature.master/ztraboo/cms-mgmt-set-advanced-fields-all-courses

mphilbrick211 commented:
Hi @ztraboo! I'm going to close this for now, but we can reopen in the future when you're able to pursue. Thanks!

openedx-webhooks commented:
@ztraboo Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/31449&entry.1484655318=https://github.com/ztraboo&entry.752974735=2022-12-14+20:34&entry.1917517419=2023-10-24+21:00&entry.2133058324=No) so we can improve your experience in the future.

<!-- comment:end_survey -->
<!-- data: eyJqaXJhLXByZS1jbG9zZSI6IG51bGx9 -->

openedx-webhooks commented:
@ztraboo Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/31449&entry.1484655318=https://github.com/ztraboo&entry.752974735=2022-12-14+20:34&entry.1917517419=2023-10-24+21:00&entry.2133058324=No) so we can improve your experience in the future.

<!-- comment:end_survey -->
<!-- data: eyJqaXJhLXByZS1jbG9zZSI6IG51bGx9 -->
